index.rst

```diff
@@ -1,61 +1,486 @@
+That metadata should be passed through a new optional keyword argument to ``Butler.put``, and can be prepared by the execution system to include provenance information.
+As implied by the recommendation text, this could only be implemented in formatters that use data formats that can store flexible metadata, but this should be true in practice for any data format used for public data products (including FITS, Parquet, and JSON).
+
+To satisfy *[REC-FIL-1]*, all we need to store is the dataset's UUID, which (after implementing the design described in [:ref:`recording-provenance`]) will be available to the execution system for insertion into the provenance metadata when the dataset is first saved to disk.
+It would be easy and possibly useful to also save the UUID of the producing quantum, and it may be worth also recording the UUIDs of all datasets input to the quantum (grouped by dataset type) as well, to possibly avoid the need for full provenance queries in the simplest cases.
+This may be less useful than it seems or perhaps even slightly confusing in some cases, however, because:
+
+- as of the time a dataset is written, we can only reliably know the "available" inputs to the quantum; the task may *later* declare that only some of these inputs were actually used;
+- in a few cases (large gather-scatter sequence points in the graph), the number of inputs to the quantum may be large (in the case of FGCM, a full-survey sequence point, it will be enormous);
+- there is no guarantee that the datasets directly input (as opposed to transitively used as inputs via some predecessor quantum's outputs) to a quantum constitute scientifically interesting provenance.
+
+At this time, it seems prudent to save only the dataset and quantum UUIDs, absent a clear use case for saving more.
+
+- *[REC-FIL-2] A study should be made of the possibility of embedding a DataLink or other service pointer in the FITS header in lieu of representing the provenance graph in the file.*
+
+The formatter hook described in the previous subsection would clearly be capable of embedding such a link in the file when it is first written, but doing so effectively declares that the service pointer can
+
+- be calculated from some combination of information (dataset UUID, data ID, dataset type, ``RUN`` collection, URI(s)) known to the execution system or formatter;
+- be assumed to remain static over the lifetime of the file.
+
+These seem like dangerous assumptions, in that satisfying them probably either creates unwanted dependencies between software components (the execution system has to be configured to know about RSP service endpoints) or puts undesirable constraints on others.
+
+Injecting this metadata into files when they are retrieved seems much cleaner conceptually, but it may rule out simple and/or efficient approaches to data access that would otherwise be on the table.
```

Comment:
We have pondered doing this type of file rewriting on get in order to support a request for HDF5 version of the dataset. Injecting FITS headers on get could work at the cost of performance (and checksums changing)